File: bin/game/PersonajeBueno.js
import { Personaje } from "./Personaje.js";
import { visionRandom } from "./Vision.js";
import { armaRandom } from "./Arma.js";

export class PersonajeBueno extends Personaje {
  constructor(position) {
    super(position, "ß", visionRandom(), armaRandom());

    const i = Math.floor(Math.random() * 8);
    this.dy = Personaje.DIRECCIONES_Y[i];
    this.dx = Personaje.DIRECCIONES_X[i];
  }

  decidirMovi(board, malos, ocupados) {
    if (!this.isEstaVivo()) return;

    let masCercano = null;
    let dist = Number.POSITIVE_INFINITY;

    for (const m of malos) {
      if (!m.isEstaVivo()) continue;
      const d = this.getPosicionActual().distanciaManhattan(m.getPosicionActual());
      if (d < dist) {
        dist = d;
        masCercano = m;
      }
    }

    const RADIO_PELIGRO = 5;

    if (masCercano && dist <= RADIO_PELIGRO) {
      for (let paso = 0; paso < 1; paso++) {
        const mejorPos = this.calcularMejorMovimiento(board, masCercano.getPosicionActual(), true, ocupados);
        if (!mejorPos) break;
        if (!this.intentarMovimiento(mejorPos, board, ocupados)) break;
      }
      return;
    }

    const UMBRAL_PATRULLA = 5;

    if (malos.length <= UMBRAL_PATRULLA) {
      const next = this.getPosicionActual().desplazar(this.dy, this.dx);

      if (!this.intentarMovimiento(next, board, ocupados)) {
        const start = Math.floor(Math.random() * 8);
        for (let k = 0; k < 8; k++) {
          const i = (start + k) % 8;
          const cand = this.getPosicionActual().desplazar(Personaje.DIRECCIONES_Y[i], Personaje.DIRECCIONES_X[i]);
          if (this.intentarMovimiento(cand, board, ocupados)) {
            this.dy = Personaje.DIRECCIONES_Y[i];
            this.dx = Personaje.DIRECCIONES_X[i];
            break;
          }
        }
      }
    }
  }
}

File: java/src/game/PersonajeBueno.java
package game;

import java.util.List;

public class PersonajeBueno extends Personaje {

    private int dy;
    private int dx;

    // Constructor del personaje bueno, asigna una posición inicial, un símbolo,
    // visión y arma aleatorias.
    public PersonajeBueno(Posicion position){
        super(position, "ß", Vision.random(), Arma.random());

        int i = (int)(Math.random() * 8);
        dy = DIRECCIONES_Y[i];
        dx = DIRECCIONES_X[i];
    }

    /**
     * Decide y mueve el personaje bueno
     * @param board
     * @param malos
     * @param ocupados
     */
    public void decidirMovi(TableroDeJuego board, List<PersonajeMalo> malos, List<Personaje> ocupados){

        if(!isEstaVivo()) return;

        PersonajeMalo masCercano = null;
        int dist = Integer.MAX_VALUE;

        // Se busca el malo más cercano dentro del radio de peligro (o cualquier malo si quedan pocos)
        for(PersonajeMalo m: malos){ //m es un malo
            if(!m.isEstaVivo()) continue;
            int d = getPosicionActual().distanciaManhattan(m.getPosicionActual());
            if(d < dist){
                dist = d;
                masCercano = m;
            }
        }

        //Cantidad de casillas que "ve" el bueno en su radio para detectar el peligro
        int RADIO_PELIGRO = 5;

        //Si entra un malo en su radio de peligro el bueno huye
        if(masCercano != null && dist <= RADIO_PELIGRO){
            for(int paso=0; paso<1; paso++){
                Posicion mejorPos = calcularMejorMovimiento(board, masCercano.getPosicionActual(), true, ocupados);
                if(mejorPos == null) break;
                if(!intentarMovimiento(mejorPos, board, ocupados)) break;
            }
            return;
        }
        
        int UMBRAL_PATRULLA = 5;

        if(malos.size() <= UMBRAL_PATRULLA){
            Posicion next = getPosicionActual().desplazar(dy, dx);

            if(!intentarMovimiento(next, board, ocupados)){
                int start = (int)(Math.random() * 8);
                for(int k=0;k<8;k++){
                    int i = (start + k) % 8;
                    Posicion cand = getPosicionActual().desplazar(DIRECCIONES_Y[i], DIRECCIONES_X[i]);
                    if(intentarMovimiento(cand, board, ocupados)){
                        dy = DIRECCIONES_Y[i];
                        dx = DIRECCIONES_X[i];
                        break;
                    }
                }
            }
        }
    }
}
